### PR TITLE
dataflow-types: make grpc clients/servers generic over the command/response types

### PR DESCRIPTION
**Pull request by guswynn**
This pr is mostly in service of making the environmentd-storaged networking also grpc.
It is mostly code movement/genericization, with some subtle points:
- The only runtime changes are changing some loglines from `GrpcComputeClient` to `GrpcClient`
- I opted for `struct GrpcClient<Client, PC, PR>` over `struct Grpc<Client: BidiProtoClient>` and using `BidiProtoClient::ProtoCommand`, I think its similar levels of clarity
- Because prost doesnt provide traits that are generic over req/response types, there is now a section in `mod grpc` where some things MUST be implemented the exact same way, though in this pr, its just the compute impls (a later pr will add the storage one) its unclear if these can ever be cleaned up.

### Motivation

   * This PR refactors existing code.

### Testing

- [X] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

None

### CODE REVIEW

**Review by benesch — APPROVED** [review]

**Review by lluki — APPROVED** [review]

File: src/dataflow-types/src/client.rs [hunk 1]
```diff
@@ -1019,123 +1024,115 @@ pub mod tcp {
...
+                        self.state = GrpcTcpConn::Disconnected;
                     }
                 }
             }
         }
     }
 
     #[async_trait]
-    impl<T> GenericClient<ComputeCommand<T>, ComputeResponse<T>> for GrpcComputedClient
+    impl<Client, C, R, PC, PR> GenericClient<C, R> for GrpcClient<Client, PC, PR>
```
Comment (antiguru) [inline src/dataflow-types/src/client.rs:1112]: We could think about making the command/response types associated types of the `GenericClient` trait, which might remove some generic parameters here and elsewhere.

File: src/dataflow-types/src/client.rs [hunk 2]
```diff
@@ -1487,43 +1504,159 @@ pub mod tcp {
...
+                Ok(resp_rx_wrap) => Ok(resp_rx_wrap.into_inner()),
+                Err(err) => Err(err)?,
+            }
+        }
+    }
+
+    // Convenient re-exports
+    pub use super::proto_compute_server::ProtoComputeServer;
+
+    // Convenientence methods for serving a grpc server that is connected to a local client
```
Comment (antiguru) [inline src/dataflow-types/src/client.rs:1593]: Spelling

File: src/dataflow-types/src/client.rs [hunk 3]
```diff
@@ -1487,43 +1504,159 @@ pub mod tcp {
...
+    pub struct ServeConfig {
+        pub listen_addr: String,
+        pub linger: bool,
```
Comment (antiguru) [inline src/dataflow-types/src/client.rs:1596]: Doc comments, please :)

File: src/dataflow-types/src/client.rs [hunk 4]
```diff
@@ -1019,123 +1024,115 @@ pub mod tcp {
     /// Streaming<ComputeResponse> instance). The recv and send functions interact with the two
     /// mpsc channel or the streaming instance respectively.
     #[derive(Debug)]
-    pub struct GrpcComputedClient {
+    pub struct GrpcClient<Client, PC, PR> {
```
Comment (antiguru) [inline src/dataflow-types/src/client.rs:1027]: The comment above needs updating as it's not specific to compute anymore.

**Review by antiguru — APPROVED** [review]
I think this looks good. Please make sure that doc comments are up-to-date and public items are documented, but otherwise this seems good to land. Thanks!

File: src/dataflow-types/src/client.rs [hunk 5]
```diff
@@ -1331,7 +1333,18 @@ pub mod tcp {
...
+    /// Constructs a framed connection for the client.
+    pub fn framed_client<A, C, R>(conn: A) -> FramedClient<A, C, R>
+    where
+        A: AsyncRead + AsyncWrite,
+    {
+        tokio_serde::Framed::new(
+            tokio_util::codec::Framed::new(conn, length_delimited_codec()),
+            Bincode::default(),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/dataflow-types/src/client.rs:1345]: we used it for storaged, I can delete it when I switch that over! I need to proto-ize the storagecommand and storageresponse first tho!

File: src/dataflow-types/src/client.rs (hunk 1 above)
Comment (guswynn) [reply src/dataflow-types/src/client.rs:1112]: ooh that would be a nice refactor!

**Pull request merged**